{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bde0fb5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "import subprocess\n",
    "genome = os.path.expanduser(\"~/clement/genomes/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9cf4848b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CHANGE-seq reveals genetic and epigenetic effects on CRISPRâ€“Cas9 genome-wide activity\n",
    "#Cicera R. Lazzarotto, Nikolay L. Malinin, Yichao Li, Ruochi Zhang, Yang Yang, GaHyun Lee, Eleanor Cowley, Yanghua He, Xin Lan, Kasey Jividen, Varun Katta, Natalia G. Kolmakova, Christopher T. Petersen, Qian Qi, Evgheni Strelcov, Samantha Maragh, Giedre Krenciute, Jian Ma, Yong Cheng & Shengdar Q. Tsai "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3f211a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CTLA4_Site9: GGACTGAGGGCCATGGACACNGG\n",
    "\n",
    "#Note that some of the targets are annotated with '|3|' in the supp_2 file, but this includes targets with 2 mismatches in the guide and 1 in the pam\n",
    "#Also, there are some with '|2|' annotations that have one mismatch and a gap\n",
    "#Here we'll only include targets with 2 mismatches and no gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7f2a18c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping target with gap: CTLA4_site_9|chr10:112098238-112098259:-|GGACTGAGGGCT-TGGACACTGG|546|2|A\tTCAGGCTTCAGCAGACCCAGACA\tGAGCCCATGAAGCAGGGTCCCAG\n",
      "wrote CTLA4_site_9|chr2:203870828-203870850:+|GGACTGAGGGCCATGGACACGGG|12382|0|both\n",
      "wrote CTLA4_site_9|chr9:135268714-135268736:+|GGACAGAGGGCCCTGGACACAGG|10036|2|both\n",
      "wrote CTLA4_site_9|chr5:101611404-101611426:+|GGAATGAGGCCCATGGACACTGG|5244|2|both\n",
      "wrote CTLA4_site_9|chr9:121888845-121888867:+|GGACTGGGGGCCTTGGACACAGG|2284|2|both\n",
      "Skipping target with gap: CTLA4_site_9|chr9:36900766-36900787:-|GGACTGGGGGC-ATGGACACAGG|162|2|C\tGGGGACCCGGGAGATAATGGAGCC\tCCAGCTCTGCTCCTCCCCATCGT\n",
      "Skipping target with gap: CTLA4_site_9|chr22:20321436-20321457:-|GGACTGAGGGC-AGGGACACTGG|70|2|C\tCTCACAGCCCCCACCCCACCTG\ttTGGTCACTGTCACCTCCCCCGG\n",
      "Skipping target with gap: CTLA4_site_9|chr22:18903849-18903870:+|GGACTGAGGGC-AGGGACACTGG|42|2|C\tGACCTGGCAGACCCCGGAGCTG\tTTGGTCACTGTCACCTCCCCCGG\n",
      "Skipping target with gap: CTLA4_site_9|chr22:45142368-45142389:-|GGCCTGAGGGCCATGGA-ACAGG|12|2|C\tTCCATGTGATACCCACCCAGGCAGG\tTGGCTGTCCCCACACAACCGCTA\n",
      "Wrote 4 targets to 01_make_demo.ipynb.genome.fa\n"
     ]
    }
   ],
   "source": [
    "targets_written = 0\n",
    "with open('../supp_2.txt','r') as fin, open('01_make_demo.ipynb.genome.fa','w') as fout, open('01_make_demo.ipynb.genome.fa.info','w') as finfo:\n",
    "    supp_line_head = fin.readline()\n",
    "    for line in fin: #CTLA4_site_9|chr16:88695984-88696006:+|GGCCTGAGGCCCTTGGACACTGG|4040|3|A GGAGGAGGCAGGAGGCAGTGACT gccaccTCAGCACCTCCTGGATT\n",
    "        if not line.strip():  # Skip empty lines\n",
    "            continue\n",
    "        line_els = line.split(\"\\t\")\n",
    "        target_els = line_els[0].split(\"|\")\n",
    "        chrom, target_range, strand = target_els[1].split(\":\")\n",
    "        num_mismatches = target_els[4]\n",
    "        if int(num_mismatches) > 2:\n",
    "            continue\n",
    "        target_seq = target_els[2]\n",
    "        if '-' in target_seq:\n",
    "            print('Skipping target with gap: ' + line.strip())\n",
    "            continue\n",
    "        if not chrom.startswith('chr'):\n",
    "            raise Exception('Cannot parse location from line ' + line.strip())\n",
    "        target_start, target_end = target_range.split(\"-\")\n",
    "        seq_lines = subprocess.check_output(['samtools', 'faidx', genome, f'{chrom}:{int(target_start)-500}-{int(target_end)+500}']).decode('utf-8')\n",
    "        seq = seq_lines[seq_lines.index('\\n')+1:].replace('\\n', '').upper()\n",
    "        fout.write(f'>CTLA4_site{targets_written}\\n{seq}\\n')\n",
    "        finfo.write(f'CTLA4_site{targets_written}\\t{chrom}:{target_start}-{target_end}:{strand}\\t{num_mismatches}\\t{line_els[0]}\\n')\n",
    "        print('wrote ' + line_els[0])\n",
    "        targets_written += 1\n",
    "print(f'Wrote {targets_written} targets to 01_make_demo.ipynb.genome.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7598c2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# link 01_make_genome.ipynb.genome.fa to test_genome.fa\n",
    "if not os.path.exists('demo_genome.fa'):\n",
    "    os.symlink('01_make_demo.ipynb.genome.fa', 'demo_genome.fa')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30c5536a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"demo_genome.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  demo_genome.fa\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1023\n",
      "Using parameters --bmax 768 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 768 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 583.714 (target: 767)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (768) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 605 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 606 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (768) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 582 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 583 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (768) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 619 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 620 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (768) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 648 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 649 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (768) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 631 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 632 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (768) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 441 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 442 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (768) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 560 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 561 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 986\n",
      "fchr[G]: 1945\n",
      "fchr[T]: 3072\n",
      "fchr[$]: 4092\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4195969 bytes to primary EBWT file: demo_genome.1.bt2.tmp\n",
      "Wrote 1028 bytes to secondary EBWT file: demo_genome.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 4092\n",
      "    bwtLen: 4093\n",
      "    sz: 1023\n",
      "    bwtSz: 1024\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 256\n",
      "    offsSz: 1024\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 22\n",
      "    numLines: 22\n",
      "    ebwtTotLen: 1408\n",
      "    ebwtTotSz: 1408\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1023\n",
      "Using parameters --bmax 768 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 768 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 3, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 583.714 (target: 767)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (768) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 640 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 641 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (768) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 303 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 304 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (768) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 631 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 632 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (768) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 614 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 615 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (768) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 671 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 672 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (768) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 589 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 590 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (768) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 638 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 639 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 986\n",
      "fchr[G]: 1945\n",
      "fchr[T]: 3072\n",
      "fchr[$]: 4092\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4195969 bytes to primary EBWT file: demo_genome.rev.1.bt2.tmp\n",
      "Wrote 1028 bytes to secondary EBWT file: demo_genome.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 4092\n",
      "    bwtLen: 4093\n",
      "    sz: 1023\n",
      "    bwtSz: 1024\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 256\n",
      "    offsSz: 1024\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 22\n",
      "    numLines: 22\n",
      "    ebwtTotLen: 1408\n",
      "    ebwtTotSz: 1408\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n",
      "Renaming demo_genome.3.bt2.tmp to demo_genome.3.bt2\n",
      "Renaming demo_genome.4.bt2.tmp to demo_genome.4.bt2\n",
      "Renaming demo_genome.1.bt2.tmp to demo_genome.1.bt2\n",
      "Renaming demo_genome.2.bt2.tmp to demo_genome.2.bt2\n",
      "Renaming demo_genome.rev.1.bt2.tmp to demo_genome.rev.1.bt2\n",
      "Renaming demo_genome.rev.2.bt2.tmp to demo_genome.rev.2.bt2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.call(['bowtie2-build', 'demo_genome.fa', 'demo_genome'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "174e5a5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Treated_1', 'Treated_2', 'Treated_3', 'Control_1', 'Control_2', 'Control_3']\n"
     ]
    }
   ],
   "source": [
    "samples = pd.read_csv('../samples.small.txt',sep=\"\\t\")\n",
    "sample_names = samples['Name'].tolist()\n",
    "for sample_name in sample_names:\n",
    "    sample_fastq = f'{sample_name}.fastq.gz'\n",
    "    if os.path.exists(sample_fastq):\n",
    "        os.remove(sample_fastq)\n",
    "print(sample_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2ba5cbb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File Treated_1.fastq.gz has 4488 lines\n",
      "File Treated_2.fastq.gz has 4100 lines\n",
      "File Treated_3.fastq.gz has 4364 lines\n",
      "File Control_1.fastq.gz has 4360 lines\n",
      "File Control_2.fastq.gz has 4440 lines\n",
      "File Control_3.fastq.gz has 4464 lines\n"
     ]
    }
   ],
   "source": [
    "\n",
    "with open(\"../CRISPRessoSea_output_on_samples.small.txt/aggregated_stats_all.txt\",'r') as fin:\n",
    "    header = fin.readline()\n",
    "    for line in fin:\n",
    "        line_els = line.split(\"\\t\")\n",
    "        target_mismatches = 5\n",
    "        match = re.search(r'_OB(\\d+)_', line_els[1])\n",
    "        if match is None:\n",
    "            if '_ON_' in line_els[1]:\n",
    "                target_mismatches = 0\n",
    "        else:\n",
    "            target_mismatches = int(match.group(1))\n",
    "        if target_mismatches > 2:\n",
    "            continue\n",
    "        region_name = line_els[12]\n",
    "        for sample_name in sample_names:\n",
    "            fastq_loc = f'../CRISPRessoSea_output_on_samples.small.txt/CRISPResso_output/CRISPRessoPooled_on_{sample_name}/MAPPED_REGIONS/REGION_{region_name}.fastq.gz'\n",
    "            if not os.path.exists(fastq_loc):\n",
    "                raise Exception('Cannot find fastq file for sample ' + sample_name + ' at ' + fastq_loc)\n",
    "            subprocess.call(f'cat {fastq_loc} >> {sample_name}.fastq.gz', shell=True)\n",
    "\n",
    "for sample_name in sample_names:\n",
    "    file_line_count = subprocess.check_output([f'zcat {sample_name}.fastq.gz | wc -l'], shell=True).decode('utf-8').strip()\n",
    "    print(f'File {sample_name}.fastq.gz has {file_line_count} lines')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f4368249",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('samples.demo.txt','w') as fout:\n",
    "    fout.write('Name\\tr1\\n')\n",
    "    for sample_name in sample_names:\n",
    "        fout.write(f'{sample_name}\\t{sample_name}.fastq.gz\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1e5b3b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "kc_crisprlungo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
